In [ ]:
import itertools

def bootstrap(M, data_list):
    bootstrap_res = itertools.product(data_list, repeat=M) 
    return bootstrap_res
# в теории именно это и надо использовать но может сильно быстро возрастать n^M
# repeat - длина того что выдаст
# а колиество будет видимо n^M то есть сразу все. поэтому я буду считать 200 лет

In [15]:
import numpy as np
rng = np.random.default_rng(seed=42)
X = np.random.exponential(scale=1.0, size=10000)
M = 10000
n = len(X)

In [19]:
bootstrap_mads = []
bootstrap_vars = [] #для выборочной дисперсии

for i in range(M):
    indices = rng.choice(n, size=n, replace=True)
    bootstrap_sample = X[indices]
    # MAD
    mad_val = np.median(np.abs(bootstrap_sample - np.median(bootstrap_sample)))
    bootstrap_mads.append(mad_val)
    # Выборочная дисперсия (смещенная - с делением на n)
    var_val = np.var(bootstrap_sample)
    bootstrap_vars.append(var_val)

bootstrap_mads.sort()
bootstrap_vars.sort()

In [23]:
# полученные значения статистики это новая выборка
# для нее получаем оценку ДИ 95%
# то есть M = 100%
# x = 2.5%
# M потому что MAD у нас M штук

two_and_a_halp_perc = int(M * 2.5 / 100)  # чилсло
lower_bound = bootstrap_mads[two_and_a_halp_perc]
upper_bound = bootstrap_mads[M - two_and_a_halp_perc - 1]

DI_MAD = bootstrap_mads[two_and_a_halp_perc : M - two_and_a_halp_perc - 1]
print(f"95% ДИ для MAD: [{lower_bound:.4f}, {upper_bound:.4f}]")
print(f"Ширина интервала: {upper_bound - lower_bound:.4f}")

lower_bound = bootstrap_vars[two_and_a_halp_perc]
upper_bound = bootstrap_vars[M - two_and_a_halp_perc - 1]
DI_VAR = bootstrap_vars[two_and_a_halp_perc : M - two_and_a_halp_perc - 1]
print(f"95% ДИ для Выборочной дисперсии: [{lower_bound:.4f}, {upper_bound:.4f}]")
print(f"Ширина интервала: {upper_bound - lower_bound:.4f}")

95% ДИ для MAD: [0.4660, 0.4921]
Ширина интервала: 0.0261
95% ДИ для Выборочной дисперсии: [0.9271, 1.0272]
Ширина интервала: 0.1001


    MAD = median(|X_i - median(X)|) - устойчив к выбросам

    Дисперсия = mean((X_i - mean(X))²) - чувствителен к выбросам

Для экспоненциального распределения:

    Данные имеют "тяжелый хвост" (много больших значений)

    MAD будет меньше подвержен влиянию этих выбросов

    Дисперсия будет сильно зависеть от выбросов

    1. Разная ширина доверительных интервалов

    MAD: ширина = 0.0261 (узкий интервал)

    Дисперсия: ширина = 0.1001 (в 3.8 раза шире!)

Вывод: Оценка дисперсии значительно менее точна, чем оценка MAD, для экспоненциального распределения.

